package/MDAnalysis/core/groups.py

```diff
@@ -1039,32 +1039,63 @@ def wrap(self, compound="atoms", center="com", box=None):
             if not all(s == 0.0):
                 o.atoms.translate(s)
 
-    def groupby(self, topattr):
+    def groupby(self, topattrs):
         """Group together items in this group according to values of *topattr*
 
         Parameters
         ----------
-        topattr: str
-           Topology attribute to group components by.
+        topattrs: str or list
+           One or more topology attribute to group components by.
+           Single arguments are passed as a string. Multiple arguments
+           are passed as a list.
 
         Returns
         -------
         dict
-            Unique values of the topology attribute as keys, Groups as values.
+            Unique values of the multiple combinations of topology attributes 
+            as keys, Groups as values.
 
         Example
         -------
-        To group atoms with the same mass together::
-
-          >>> ag.groupby('masses')
-          {12.010999999999999: <AtomGroup with 462 atoms>,
-          14.007: <AtomGroup with 116 atoms>,
-          15.999000000000001: <AtomGroup with 134 atoms>}
-
+       
+        
+        To group atoms with the same mass together::     
+                
+        >>> ag.groupby('masses')
+        {12.010999999999999: <AtomGroup with 462 atoms>,
+         14.007: <AtomGroup with 116 atoms>,
+         15.999000000000001: <AtomGroup with 134 atoms>}        
+
+        To group atoms with the same residue name and mass together::
+
+          >>> ag.groupby(['resnames', 'masses'])
+          {'ALA': {1.008: <AtomGroup with 95 atoms>,
+            12.011: <AtomGroup with 57 atoms>,
+            14.007: <AtomGroup with 19 atoms>,
+            15.999: <AtomGroup with 19 atoms>},
+           'ARG': {1.008: <AtomGroup with 169 atoms>,
+           ...}
+          
+          >>> ag.groupby(['resnames', 'masses'])['ALA'][15.999]
+           <AtomGroup with 19 atoms>
+        
         .. versionadded:: 0.16.0
+        .. versionchanged:: 0.18.0 The function accepts multiple attributes
         """
-        ta = getattr(self, topattr)
-        return {i: self[ta == i] for i in set(ta)}
+        
+        res = {}
+        if type(topattrs) == str:
+            ta = getattr(self, topattrs)
+            return {i: self[ta == i] for i in set(ta)}
+        else:
+            ta=getattr(self,topattrs[0])
+            for i in set(ta):
+                if len(topattrs) == 1:
+                    res[i] = self[ta == i]
+                else:
+                    res[i] = self[ta == i].groupby(topattrs[1:])
+            return res
```

Comment:
Return a dict with keys as tuples; see perhaps https://stackoverflow.com/a/6027615 for a general `flatten()` function that can be adapted to produce key tuples.

Reply:
See https://github.com/MDAnalysis/mdanalysis/pull/1846#issuecomment-376699783 for a general `flatten()` that creates tuple keys.

If that version of `flatten()` works then you can put it into `lib.utils.flatten_dict()` or similar. Add as a test case the example that I used.

Reply:
@orbeckst Thank you for your help.

I got this function working to flatten the output of groupby
```python
    import collections
    def flatten_dict(d, parent_key=''):
        items = []
        for k, v in d.items():
            new_key = (parent_key, k) if parent_key else k
            if isinstance(v, collections.MutableMapping):
                 items.extend(flatten_dict(v, new_key).items())
            else:
            items.append((new_key, v))
            return dict(items)
```

Reply:
@orbeckst Somehow I missed the edit to your comment https://github.com/MDAnalysis/mdanalysis/pull/1846#issuecomment-376699783

Reply:
submit a PR and ping me and I will review.

Reply:
... "submit a PR" – I meant, just push a new commit with the changes.

Reply:
@orbeckst I tried your code, but it fails on this line:
``` new_key = parent_key+tuple(k)
TypeError: 'numpy.float32' object is not iterable
```
while using (k, ) instead works fine

Reply:
Use what works :-)